internal/tasks/consumer.go

```diff
@@ -33,6 +34,8 @@ func dispatchTasks(store store.Store, callbackManager CallbackManager, k8sClient
 			return deviceRender(ctx, &reference, store, callbackManager, k8sClient, configStorage, log)
 		case RepositoryUpdatesTask:
 			return repositoryUpdate(ctx, &reference, store, callbackManager, log)
+		case SignerTask:
+			return asyncSign(ctx, &reference, store, callbackManager, log, ca)
```

Comment:
Retry revolves around internal logic in casign.go/asyncSign and friends. 

They will process all outstanding requests in the database. If a new request is added in the meantime it will be processed as well. This allows any number of worker threads as well scaling them horizontally.

The worker continues to process until there are zero candidates in the database. 

Unfortunately, the whole thing is prone to a race. There is a window if the record is added after the "get next records" query, but before the processing loop returns and releases the lock. It is not large and it is very unlikely because this means that insertion, submission of publish request, receiving it and handling has happened faster than for _, ereq := range ereqs.Items with len(Items) == 0. 

I will be honest - I do not know how to eliminate this one short of having a "cronjob" to do a regular re-run.

Reply:
To be clear - the race can happen only if there are ZERO records to be processed in the database. If there is even one, adding other while it is being processed should be handled correctly.